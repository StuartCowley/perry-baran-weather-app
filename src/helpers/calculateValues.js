const calcMean = (array, param, decimals = 1) => {
  return (
    array.reduce((prev, curr) => {
      return prev + curr[param];
    }, 0) / array.length
  ).toFixed(decimals);
};

const calcMax = (array, param, decimals = 1) => {
  const values = array.map((item) => {
    return item[param];
  });
  return Math.max(...values).toFixed(decimals);
};

const calcMin = (array, param, decimals = 1) => {
  const values = array.map((item) => {
    return item[param];
  });
  return Math.min(...values).toFixed(decimals);
};

export { calcMean, calcMax, calcMin };
