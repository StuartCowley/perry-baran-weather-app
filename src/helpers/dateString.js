const dateString = (dateTime) => {
  return new Date(dateTime * 1000).toDateString();
};

export default dateString;
