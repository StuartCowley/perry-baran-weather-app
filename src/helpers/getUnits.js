const getUnits = (units) => {
  let selectedUnits;
  if (typeof units === "string") selectedUnits = units.toLowerCase();
  switch (selectedUnits) {
    case "imperial":
      return {
        tempUnits: "°F",
        speedUnits: "mph",
      };
    case "metric":
      return {
        tempUnits: "°C",
        speedUnits: "m/s",
      };
    case "standard":
    default:
      return {
        tempUnits: "K",
        speedUnits: "m/s",
      };
  }
};

export default getUnits;
