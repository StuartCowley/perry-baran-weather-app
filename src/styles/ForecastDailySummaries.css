.forecast-daily-summaries {
  display: flex;
  justify-content: space-around;
}