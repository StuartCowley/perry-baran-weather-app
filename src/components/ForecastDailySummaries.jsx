import React from "react";
import "../styles/ForecastDailySummaries.css";
import PropTypes from "prop-types";
import ForecastSummary from "./ForecastDailySummary";
import { calcMean } from "../helpers/calculateValues";

function ForecastDailySummaries({
  forecasts,
  handleForecastSelect,
  selectedDate,
}) {
  const getDailyForecast = (forecastsArray) => {
    return forecastsArray.map((forecast) => {
      const [{ dateTime, icon, weather }] = forecast;
      return {
        dateTime,
        icon,
        temp: calcMean(forecast, "temp"),
        weather,
      };
    });
  };

  const dailyForecast = getDailyForecast(forecasts);

  return (
    <div className="forecast-daily-summaries">
      {dailyForecast.map((forecast) => {
        const { dateTime, weather, icon, temp } = forecast;
        return (
          <ForecastSummary
            dateTime={dateTime}
            weather={weather}
            icon={icon}
            temp={temp}
            handleForecastSelect={handleForecastSelect}
            selectedDate={selectedDate}
            key={forecast.dateTime}
          />
        );
      })}
    </div>
  );
}

ForecastDailySummaries.defaultProps = {
  selectedDate: 0,
};

ForecastDailySummaries.propTypes = {
  forecasts: PropTypes.arrayOf(
    PropTypes.arrayOf(
      PropTypes.shape({
        dateTime: PropTypes.number.isRequired,
        icon: PropTypes.string.isRequired,
        temp: PropTypes.number.isRequired,
        weather: PropTypes.string.isRequired,
      })
    )
  ).isRequired,
  handleForecastSelect: PropTypes.func.isRequired,
  selectedDate: PropTypes.number,
};

export default ForecastDailySummaries;
