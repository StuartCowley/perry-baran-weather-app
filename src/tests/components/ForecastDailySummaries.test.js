import React from "react";
import { render } from "@testing-library/react";
import ForecastDailySummaries from "../../components/ForecastDailySummaries";
import { calcMean } from "../../helpers/calculateValues";
import dateString from "../../helpers/dateString";

describe("ForecastDailySummaries", () => {
  const validProps = {
    forecasts: [
      [
        {
          dateTime: 1667487600,
          icon: "200",
          temp: 10,
          weather: "unique1 weather1",
        },
        {
          dateTime: 2667554800,
          icon: "201",
          temp: 11,
          weather: "unique2 weather2",
        },
      ],
      [
        {
          dateTime: 1667898000,
          icon: "800",
          temp: 12,
          weather: "unique3 weather3",
        },
        {
          dateTime: 2667865600,
          icon: "801",
          temp: 13,
          weather: "unique4 weather4",
        },
      ],
      [
        {
          dateTime: 1677447600,
          icon: "600",
          temp: 14,
          weather: "unique5 weather5",
        },
        {
          dateTime: 2677554800,
          icon: "601",
          temp: 15,
          weather: "unique6 weather6",
        },
      ],
    ],
    handleForecastSelect: () => {},
  };
  const { forecasts, handleForecastSelect } = validProps;
  let screen;

  beforeEach(() => {
    screen = render(
      <ForecastDailySummaries
        forecasts={forecasts}
        handleForecastSelect={handleForecastSelect}
      />
    );
  });

  test("snapshot", () => {
    const { asFragment } = screen;

    expect(asFragment()).toMatchSnapshot();
  });

  test("ForecastDailySummary is called correct amount of times", () => {
    const { getAllByTestId } = screen;

    expect(getAllByTestId("forecast-daily-summary")).toHaveLength(
      forecasts.length
    );
  });

  test("data is properly manipulated", () => {
    const { queryByText } = screen;

    forecasts.forEach((forecast) => {
      const temp = calcMean(forecast, "temp");
      const date0 = dateString(forecast[0].dateTime).slice(3, 10);
      const description0 = forecast[0].weather.split(" ")[0];
      const date1 = dateString(forecast[1].dateTime).slice(3, 10);
      const description1 = forecast[1].weather.split(" ")[0];

      expect(queryByText(new RegExp(temp))).toBeInTheDocument();
      expect(queryByText(new RegExp(date0))).toBeInTheDocument();
      expect(queryByText(new RegExp(description0))).toBeInTheDocument();
      expect(queryByText(new RegExp(date1))).not.toBeInTheDocument();
      expect(queryByText(new RegExp(description1))).not.toBeInTheDocument();
    });
  });
});
