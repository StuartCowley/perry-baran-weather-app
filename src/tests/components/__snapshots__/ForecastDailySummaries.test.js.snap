// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ForecastDailySummaries snapshot 1`] = `
<DocumentFragment>
  <div
    class="forecast-daily-summaries"
  >
    <div
      class="forecast-daily-summary"
      data-testid="forecast-daily-summary"
    >
      <p>
        Thu 
        <br />
         Nov 03
      </p>
      <div
        class="forecast-daily-summary__icon"
        data-testid="forecast-icon"
      >
        <i
          class="wi wi-day-storm-showers"
          name="owm"
        />
      </div>
      <p>
        unique1 
        <br />
        weather1
      </p>
      <p>
        10.5°C
      </p>
      <button
        class="forecast-daily-summary__more-details"
        type="button"
      >
        More Details
      </button>
    </div>
    <div
      class="forecast-daily-summary"
      data-testid="forecast-daily-summary"
    >
      <p>
        Tue 
        <br />
         Nov 08
      </p>
      <div
        class="forecast-daily-summary__icon"
        data-testid="forecast-icon"
      >
        <i
          class="wi wi-day-sunny"
          name="owm"
        />
      </div>
      <p>
        unique3 
        <br />
        weather3
      </p>
      <p>
        12.5°C
      </p>
      <button
        class="forecast-daily-summary__more-details"
        type="button"
      >
        More Details
      </button>
    </div>
    <div
      class="forecast-daily-summary"
      data-testid="forecast-daily-summary"
    >
      <p>
        Sun 
        <br />
         Feb 26
      </p>
      <div
        class="forecast-daily-summary__icon"
        data-testid="forecast-icon"
      >
        <i
          class="wi wi-day-snow"
          name="owm"
        />
      </div>
      <p>
        unique5 
        <br />
        weather5
      </p>
      <p>
        14.5°C
      </p>
      <button
        class="forecast-daily-summary__more-details"
        type="button"
      >
        More Details
      </button>
    </div>
  </div>
</DocumentFragment>
`;
